engine/gamesys/src/gamesys/components/comp_particlefx.cpp

```diff
@@ -330,23 +331,42 @@ namespace dmGameSystem
                 pfx_world->m_DT, emitter_render_data->m_Instance, emitter_render_data->m_EmitterIndex,
                 emitter_attribute_info, Vector4(1,1,1,1), (void*) vertex_buffer.Begin(), vb_max_size, &vb_size);
 
-            if (res != dmParticle::GENERATE_VERTEX_DATA_OK)
+            if (res == dmParticle::GENERATE_VERTEX_DATA_MAX_PARTICLES_EXCEEDED)
             {
-                if (res == dmParticle::GENERATE_VERTEX_DATA_MAX_PARTICLES_EXCEEDED)
-                {
-                    dmLogWarning("Maximum number of particles (%d) exceeded, particles will not be rendered. Change \"%s\" in the config file.",
-                        pfx_context->m_MaxParticleCount, dmParticle::MAX_PARTICLE_COUNT_KEY);
-                }
-                else if (res == dmParticle::GENERATE_VERTEX_DATA_INVALID_INSTANCE)
-                {
-                    dmLogWarning("Cannot generate vertex data for emitter (%d), particle instance handle is invalid.", (*i));
+                if (!vb_offset) { // need to set the size of the buffer
+                    uint32_t vx_size = 0;
+                    for (uint32_t *t = begin; t != end; ++t)
+                    {
+                        const dmParticle::EmitterRenderData* ed = (dmParticle::EmitterRenderData*) buf[*t].m_UserData;
+                        vx_size += GetEmitterVertexCount(pfx_world->m_ParticleContext, ed->m_Instance, ed->m_EmitterIndex);
+                    }
+                    vx_size *= 6 * material_attribute_info.m_VertexStride;
+                    dmRender::SetBufferData(render_context, pfx_world->m_VertexBuffer, vx_size, 0, dmGraphics::BUFFER_USAGE_STREAM_DRAW);
                 }
+
+                dmRender::SetBufferSubData(render_context, pfx_world->m_VertexBuffer, vb_offset, vb_size, vertex_buffer.Begin());
+                vertex_buffer.SetSize(0); // consumed
+                vb_offset += (vb_size - vb_size_init);
+                vb_size = vb_size_init;
+            }
+            else if (res == dmParticle::GENERATE_VERTEX_DATA_INVALID_INSTANCE)
+            {
+                dmLogWarning("Cannot generate vertex data for emitter (%d), particle instance handle is invalid.", (*i));
             }
         }
 
-        uint32_t ro_vertex_count = (vb_size - vb_size_init) / material_attribute_info.m_VertexStride;
-
-        vertex_buffer.SetSize(vb_size);
+        uint32_t ro_vertex_count;
+        if (vb_offset) { // consume the last chunk of data
+            if (vb_size) {
+                dmRender::SetBufferSubData(render_context, pfx_world->m_VertexBuffer, vb_offset, vb_size, vertex_buffer.Begin());
+                vb_offset += (vb_size - vb_size_init);
+            }
+            ro_vertex_count = vb_offset / material_attribute_info.m_VertexStride;
+            vertex_buffer.SetSize(0);
+        } else {
+            ro_vertex_count = (vb_size - vb_size_init) / material_attribute_info.m_VertexStride;
+            vertex_buffer.SetSize(vb_size);
```

Comment:
At this last stage, the vertex count should be possible to be calculated in the same fashion?

Also, I don't really understand why the buffer is set to two different sizes (0 vs vb_size)?

Reply:
In the top case the data has already been uploaded in batches so should not be done after.